log/log.go

```diff
@@ -0,0 +1,199 @@
+	}
+}
+
+func (l *logger) New(keyvals ...interface{}) Logger {
+	var kv []interface{}
+	kv = append(kv, l.ctx...)
+	kv = append(kv, expandKeyVals(keyvals)...)
+
+	return &logger{
+		ctx: kv,
+		h:   l.h,
+	}
+}
+
+func (l *logger) SetHandler(h Handler) {
+	l.h = h
+}
+
+func (l *logger) Log(keyvals ...interface{}) {
+	var kv []interface{}
+	kv = append(kv, l.ctx...)
+	// kv = append(keyvals, CalllerKey, stack.Caller(1))
+	kv = append(kv, expandKeyVals(keyvals)...)
+	l.h.Handle(kv...)
+}
+
+func expandKeyVals(keyvals []interface{}) []interface{} {
+	kvCount := len(keyvals)
+	for _, kv := range keyvals {
+		if _, ok := kv.(KeyVal); ok {
+			kvCount++
+		}
+	}
+	if kvCount == len(keyvals) {
+		return keyvals
+	}
+	exp := make([]interface{}, 0, kvCount)
+	for _, kv := range keyvals {
+		switch kv := kv.(type) {
+		case KeyVal:
+			exp = append(exp, kv.Key())
+			exp = append(exp, kv.Value())
+		default:
+			exp = append(exp, kv)
+		}
+	}
+	return exp
+}
+
+type KeyVal interface {
+	Key() interface{}
+	Value() interface{}
+}
+
+type key int
+
+const (
+	TimeKey key = iota
+	LvlKey
+	CalllerKey
+)
+
+func Now() KeyVal {
+	return logtime(time.Now())
+}
+
+type logtime time.Time
+
+func (t logtime) Key() interface{}   { return TimeKey }
+func (t logtime) Value() interface{} { return time.Time(t) }
+
+type Lvl int
+
+const (
+	Crit Lvl = iota
+	Error
+	Warn
+	Info
+	Debug
+)
+
+func (l Lvl) Key() interface{}   { return LvlKey }
+func (l Lvl) Value() interface{} { return l }
+
+func (l Lvl) MarshalText() (text []byte, err error) {
+	switch l {
+	case Crit:
+		return []byte("crit"), nil
+	case Error:
+		return []byte("eror"), nil
+	case Warn:
+		return []byte("warn"), nil
+	case Info:
+		return []byte("info"), nil
+	case Debug:
+		return []byte("dbug"), nil
+	}
+	panic("unexpected level value")
+}
+
+type Handler interface {
+	Handle(keyvals ...interface{}) error
+}
+
+type HandlerFunc func(keyvals ...interface{}) error
+
+func (f HandlerFunc) Handle(keyvals ...interface{}) error {
+	return f(keyvals...)
+}
+
+type Encoder interface {
+	Encode(keyvals ...interface{}) ([]byte, error)
+}
+
+func AddTimestamp(h Handler) Handler {
+	return HandlerFunc(func(keyvals ...interface{}) error {
+		kvs := append(keyvals, nil, nil)
+		copy(kvs[2:], kvs[:len(kvs)-2])
+		kvs[0] = TimeKey
+		kvs[1] = time.Now()
+		return h.Handle(kvs...)
+	})
+}
+
+func ReplaceKeys(h Handler, keypairs ...interface{}) Handler {
+	m := make(map[interface{}]interface{}, len(keypairs)/2)
+	for i := 0; i < len(keypairs); i += 2 {
+		m[keypairs[i]] = keypairs[i+1]
+	}
+	return HandlerFunc(func(keyvals ...interface{}) error {
+		for i := 0; i < len(keyvals); i += 2 {
+			if nk, ok := m[keyvals[i]]; ok {
+				keyvals[i] = nk
+			}
+		}
+		return h.Handle(keyvals...)
+	})
+}
+
+func Writer(w io.Writer, enc Encoder) Handler {
+	return HandlerFunc(func(keyvals ...interface{}) error {
+		b, err := enc.Encode(keyvals...)
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(b)
```

Comment:
You are correct. I left it out in this early API sketch. We could lock a mutex around the call to w.Write.

Another approach is to give options to the app code by providing a SyncHandler and an AsyncHandler. Both would wrap an arbitrary Handler. SyncHandler protects the wrapped handler via a Mutex. AsyncHandler serializes the pipeline via a (maybe buffered) channel and introduces a goroutine to run the downstream handlers asynchronously.